packages/swingset-liveslots/src/virtualObjectManager.js

```diff
@@ -586,273 +661,282 @@ export function makeVirtualObjectManager(
-          thisfulMethods,
-          interfaceGuard,
-        );
         break;
       }
       case 'not': {
         throw Fail`invalid behavior specifier for ${q(tag)}`;
       }
       default: {
-        throw Fail`unexepected facetiousness: ${q(facetiousness)}`;
+        throw Fail`invalid facetiousness`;
       }
     }
+    // beyond this point, we use 'multifaceted' to switch modes
 
-    if (durableKindDescriptor) {
-      const { unfaceted, facets } = durableKindDescriptor;
-      if (multifaceted) {
-        assert(
-          !unfaceted,
-          `durable kind "${tag}" originally defined as single-faceted`,
-        );
-        if (facets) {
-          const m = `durable kind "${tag}" facets don't match original definition`;
-          assert(
-            facetNames !== undefined && facets.length === facetNames.length,
-            m,
-          );
-          for (const [idx, facet] of facetNames.entries()) {
-            assert(facet === facets[idx], m);
-          }
-        } else {
-          durableKindDescriptor.facets = facetNames;
-          saveDurableKindDescriptor(durableKindDescriptor);
-        }
-      } else {
-        assert(
-          !facets,
-          `durable kind "${tag}" originally defined as multi-faceted`,
-        );
-        if (!unfaceted) {
-          durableKindDescriptor.unfaceted = true;
-          saveDurableKindDescriptor(durableKindDescriptor);
-        }
-      }
+    if (isDurable) {
+      checkAndUpdateFacetiousness(
+        tag,
+        durableKindDescriptor,
+        facetNames,
+        saveDurableKindDescriptor,
+      );
     }
 
-    vrm.registerKind(kindID, reanimate, deleteStoredVO, isDurable);
-    vrm.rememberFacetNames(kindID, facetNames);
-    harden(contextMapTemplate);
-    harden(prototypeTemplate);
-
-    function makeRepresentative(innerSelf, initializing) {
-      innerSelf.repCount === 0 ||
-        Fail`${innerSelf.baseRef} already has a representative`;
-      innerSelf.repCount += 1;
+    harden(stateShape);
+    stateShape === undefined ||
+      passStyleOf(stateShape) === 'copyRecord' ||
+      assert.fail(X`A stateShape must be a copyRecord: ${q(stateShape)}`);
+    assertPattern(stateShape);
 
-      function ensureState() {
-        if (innerSelf.rawState) {
-          cache.refresh(innerSelf);
-        } else {
-          innerSelf = cache.lookup(innerSelf.baseRef, true);
-        }
-      }
+    let isStateProperty = _prop => undefined;
+    /** @type {(value: any, prop: string) => void} */
+    let checkStateProperty = (_value, _prop) => undefined;
+    if (stateShape) {
+      isStateProperty = prop =>
+        hasOwnPropertyOf(stateShape, prop) ||
+        assert.fail(
+          X`State must only have fields described by stateShape: ${q(
+            ownKeys(stateShape),
+          )}`,
+        );
+      checkStateProperty = (value, prop) => {
+        isStateProperty(prop);
+        mustMatch(value, stateShape[prop]);
+      };
+    }
```

Comment:
Suggested renames (`checkStateProperty` → `checkStatePropertyValue` and `isStateProperty` → `checkStateProperty`) and adoption of `Fail`:
```suggestion
    let checkStateProperty = _prop => undefined;
    /** @type {(value: any, prop: string) => void} */
    let checkStatePropertyValue = (_value, _prop) => undefined;
    if (stateShape) {
      checkStateProperty = prop =>
        hasOwnPropertyOf(stateShape, prop) ||
        Fail`State must only have fields described by stateShape: ${q(
          ownKeys(stateShape),
        )}`;
      checkStatePropertyValue = (value, prop) => {
        checkStateProperty(prop);
        mustMatch(value, stateShape[prop]);
      };
    }
```